feat: Add user management CLI commands

manoharan-nexthop opened the pull request:
#### What I did
This implementation addresses the User Management HLD requirements for user administration in SONiC. https://github.com/sonic-net/SONiC/pull/2018

Add CLI interface for SONiC local user management:

- config user add/modify/delete - User lifecycle management
- config user add-ssh-key/remove-ssh-key - SSH key management
- config user import-existing - Import system users to SONiC
- show user [username] - Display user information and status
- show user roles - Display available roles and permissions


#### How I did it
Features:
- Role-based access control (administrator/operator)
- Secure password handling with confirmation prompts
- SSH key format validation and management
- CONFIG_DB integration for persistent configuration
- Comprehensive input validation and error handling
- Complete test suite with unit and integration tests

Commands integrate with userd daemon for system-level user operations.

#### How to verify it
Validated the CLI and its functionality.

#### Previous command output (if the output of a command-line utility has changed)

#### New command output (if the output of a command-line utility has changed)

mssonicbld commented:
/azp run

aidan-gallagher commented:
I think you have accidentally pushed your code for sonic-host-services instead of sonic-utilities with your most recent force push (from [c4ad7ec](https://github.com/sonic-net/sonic-utilities/commit/c4ad7ecca168e62e1ffa6aba4eecfcaf1b71e2ce) to [c1b2c4b](https://github.com/sonic-net/sonic-utilities/commit/c1b2c4bb3e7690b8772490ad3a0c9677c907ca6d)).

mssonicbld commented:
/azp run

aidan-gallagher commented on config/user.py:
```diff
@@ -0,0 +1,668 @@
+        config_db: Database connection
+        username: Username to update
+        user_data: User data to set
+        operation: Operation description for success message (e.g., "added", "modified")
```
nit: Operation description for **error** message

manoharan-nexthop replied:
Updated.

mssonicbld commented:
/azp run

spandan-nexthop reviewed (commented):
LGTM

manoharan-nexthop committed:
[sonic-utilities] Add user management CLI commands

Add comprehensive CLI interface for SONiC local user management:

- config user add/modify/delete - User lifecycle management
- config user add-ssh-key/remove-ssh-key - SSH key management
- config user import-existing - Import system users to SONiC
- show user [username] - Display user information and status
- show user roles - Display available roles and permissions

Features:
- Role-based access control (administrator/operator)
- Secure password handling with confirmation prompts
- SSH key format validation and management
- CONFIG_DB integration for persistent configuration
- Comprehensive input validation and error handling
- Complete test suite with unit and integration tests

Signed-off-by: Manoharan Sundaramoorthy <manoharan@nexthop.ai>

mssonicbld commented:
/azp run

qiluo-msft commented:
/Semgrep

manoharan-nexthop committed:
Handle boolean as string attribute well for user enable

CONFIG_DB stores boolean values as strings ("true"/"false"), but the
local user management code was using these values directly in Python
boolean contexts. Since any non-empty string (including "false") is
truthy in Python, this caused:
1. `config user modify --disabled` allowing ALL admin users to be
disabled
2. `show user list` displays `"Enabled: Yes"` for disabled users

Added is_true() helper function in `utilities_common/general.py` to
properly parse CONFIG_DB string booleans and used the same in
`config/user.py` and `show/user.py`

Updated the tests to verify the same and manually verified if the
configuration and show are as expected.

Signed-off-by: Manoharan Sundaramoorthy <manoharan@nexthop.ai>

ronan-nexthop commented:
@aidan-gallagher could you take another look at this, we will address the failing CI + merge conflicts

manoharan-nexthop commented on config/user.py:
```diff
@@ -0,0 +1,697 @@
+# Constants
+LOCAL_USER_TABLE = "LOCAL_USER"
+LOCAL_ROLE_SECURITY_POLICY_TABLE = "LOCAL_ROLE_SECURITY_POLICY"
+DEVICE_METADATA_TABLE = "DEVICE_METADATA"
+DEVICE_METADATA_LOCALHOST_KEY = "localhost"
+LOCAL_USER_MANAGEMENT_FIELD = "local_user_management"
+
+# System users to exclude from import
+SYSTEM_USERS = {'root', 'daemon', 'bin', 'sys', 'sync', 'games', 'man', 'lp', 'mail',
+                'news', 'uucp', 'proxy', 'www-data', 'backup', 'list', 'irc', 'gnats',
+                'nobody', '_apt', 'systemd-network', 'systemd-resolve', 'messagebus',
+                'systemd-timesync', 'sshd', 'redis', 'ntp', 'frr', 'snmp'}
```
These are system users that should be excluded and comment is is there at the top.

manoharan-nexthop commented on config/user.py:
```diff
@@ -0,0 +1,697 @@
+            # Check for empty password
+            if not password or password.strip() == "":
+                click.echo("Error: Password cannot be empty")
+                time.sleep(3)  # Security delay to prevent rapid brute-force attempts
```
Removed, this is not required.

manoharan-nexthop commented on config/user.py:
```diff
@@ -0,0 +1,697 @@
+                return None, False
+        finally:
+            # Always clear passwords from memory
+            password = None
```
This is to cleanup the memory.

manoharan-nexthop commented on config/user.py:
```diff
@@ -0,0 +1,697 @@
+        finally:
+            # Always clear passwords from memory
+            password = None
+            confirm_password = None
```
same as above.

manoharan-nexthop commented on config/user.py:
```diff
@@ -0,0 +1,697 @@
+            try:
+                policies[field] = int(policies[field])
+            except (ValueError, TypeError):
+                pass
```
Fixed.

manoharan-nexthop commented on config/user.py:
```diff
@@ -0,0 +1,697 @@
+    errors = []
+
+    # Check minimum length
+    if 'len_min' in policies:
+        min_len = policies['len_min']
+        if len(password) < min_len:
+            errors.append(f"Password must be at least {min_len} characters long")
```
Updated to handle minium length type.

manoharan-nexthop commented on config/user.py:
```diff
@@ -0,0 +1,697 @@
+                    shadow_passwords[parts[0]] = parts[1]
+    except (IOError, OSError):
+        # Shadow file may not be readable, continue with default password hash
+        pass
```
The config commands can be run only by sudoers.

manoharan-nexthop commented on config/user.py:
```diff
@@ -0,0 +1,697 @@
+                for line in f:
+                    line = line.strip()
+                    if line and not line.startswith('#'):
+                        ssh_keys.append(line)
```
Updated.

manoharan-nexthop commented on config/user.py:
```diff
@@ -0,0 +1,697 @@
+        return
+
+    # Handle password input and validation
+    password_hash, success = handle_password_input(username, password_hash,
+                                                   password_prompt, db, is_new_user=True)
+    if not success:
+        return
+
+    if not password_hash:
+        click.echo("Error: Failed to generate password hash")
+        return
```
have updated to set it to `!` when not provided.

manoharan-nexthop commented on config/user.py:
```diff
@@ -0,0 +1,697 @@
+            updated_data['ssh_keys'] = ssh_keys_result
+        elif 'ssh_keys' in updated_data:
+            # Remove ssh_keys field if result is empty (all keys were removed)
+            del updated_data['ssh_keys']
```
Changed to set it to `[]`

manoharan-nexthop commented on config/user.py:
```diff
@@ -0,0 +1,697 @@
+    config_db = get_config_db(db)
+
+    policy_data = {
+        'max_login_attempts': max_login_attempts
```
Updated.

manoharan-nexthop commented on show/user.py:
```diff
@@ -0,0 +1,239 @@
+    users = get_user_database()
+
+    user_data = users.get(current_username)
+    return user_data and user_data.get('role') == 'administrator'
```
Updated to handle the exception.

manoharan-nexthop commented on show/user.py:
```diff
@@ -0,0 +1,239 @@
+
+    # Handle both array format (real system) and string format (test system)
+    if isinstance(ssh_keys, list):
+        key_list = ssh_keys
+    elif isinstance(ssh_keys, str):
+        # Check if it's a string representation of an array (from mock Redis)
+        if ssh_keys.startswith('[') and ssh_keys.endswith(']'):
+            if ssh_keys == '[]':
+                key_list = []
+            else:
+                # Parse array string representation (simplified approach)
+                inner = ssh_keys[1:-1].strip()
+                if not inner:
+                    key_list = []
+                else:
+                    # Split on comma and clean up quotes
+                    items = [item.strip() for item in inner.split(',') if item.strip()]
+                    key_list = []
+                    for item in items:
+                        # Remove surrounding quotes if present
+                        item = item.strip()
+                        if ((item.startswith("'") and item.endswith("'")) or
+                                (item.startswith('"') and item.endswith('"'))):
+                            item = item[1:-1]
+                        key_list.append(item)
+        else:
+            # Split by comma and filter out empty strings (regular string format)
+            key_list = [key.strip() for key in ssh_keys.split(',') if key.strip()]
+    else:
+        return "  SSH Keys: None"
+
+    if not key_list:
+        return "  SSH Keys: None"
+
+    result = [f"  SSH Keys ({len(key_list)}):"]
+
+    for i, key in enumerate(key_list, 1):
+        result.append(f"    {i}. {key.strip()}")
+
+    return "\n".join(result)
```
Its added so that incase the database used in either format to work.

manoharan-nexthop commented on show/user.py:
```diff
@@ -0,0 +1,239 @@
+    click.echo(f"User: {username}")
+    click.echo(f"  Role: {user_data.get('role', 'N/A')}")
+    click.echo(f"  Enabled: {'Yes' if user_data.get('enabled', True) else 'No'}")
+
```
updated.

manoharan-nexthop commented on tests/config_user_test.py:
```diff
@@ -0,0 +1,1687 @@
+        # Verify user was NOT created due to policy validation failure
+        user_data = self.db.cfgdb.get_entry("LOCAL_USER", "edgeuser")
+        self.assertIn(user_data, [None, {}],
+                      "User should not be created when policy validation fails")
```
Updated the asserts.

manoharan-nexthop commented on tests/config_user_test.py:
```diff
@@ -0,0 +1,1687 @@
+        # Verify command succeeded
+        self.assertEqual(result.exit_code, 0)
+        # Silent on success - no output expected
+
```
updated the assert.

manoharan-nexthop commented on tests/show_user_test.py:
```diff
@@ -0,0 +1,862 @@
+        mock_get_policies.return_value = mock_policies
+
+        # Use valid role name instead of 'test_role'
+        result = self.runner.invoke(show_user.show_security_policy, ['administrator'], obj=self.db)
```
Removed.

manoharan-nexthop commented on tests/show_user_test.py:
```diff
@@ -0,0 +1,862 @@
+                result = self.runner.invoke(show_user.show_security_policy, [], obj=self.db)
+                # Should handle gracefully
+                self.assertEqual(result.exit_code, 0)
+            except Exception:
```
Changed to assert.

manoharan-nexthop commented on tests/show_user_test.py:
```diff
@@ -0,0 +1,862 @@
+                # Should handle gracefully
+                self.assertEqual(result.exit_code, 0)
+            except Exception:
+                pass
```
Updated.

manoharan-nexthop commented on tests/show_user_test.py:
```diff
@@ -0,0 +1,862 @@
+
+        result = self.runner.invoke(show_user.show_users, [], obj=self.db)
+
+        # Should handle corrupted data gracefully (may fail with exit code 1)
```
Changed the assert.

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

aidan-gallagher commented on config/user.py:
```diff
@@ -0,0 +1,699 @@
+        click.echo(f"Error: User '{username}' does not exist")
+        return
+
+    # Check admin constraint only when feature is enabled
```
>   `# (when disabled, allow deleting even the last admin since userd won't act on it)`

Is it an issue when the feature is re-enabled?

What happens when:
1. User disables user management feature
2. User deletes all admin accounts
3. User re-enables user management feature

manoharan-nexthop replied:
In the yang, we have a must condition that atleast one admin admin user has to exist. That should take care of the sequence that you have mentioned.

aidan-gallagher commented:
> @aidan-gallagher could you take another look at this, we will address the failing CI + merge conflicts

Added 1 small comment but LGTM.

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run
